erc/ERCS/erc-4337.md

```diff
@@ -434,6 +434,16 @@ Otherwise, attackers may be able to use the banned opcodes to detect running on-
 
 When a bundler includes a bundle in a block it must ensure that earlier transactions in the block don't make any UserOperation fail. It should either use access lists to prevent conflicts, or place the bundle as the first transaction in the block.
 
+
+## Rationale
+
+The main challenge with a purely smart contract wallet based account abstraction system is DoS safety: how can a block builder including an operation make sure that it will actually pay fees, without having to first execute the entire operation? Requiring the block builder to execute the entire operation opens a DoS attack vector, as an attacker could easily send many operations that pretend to pay a fee but then revert at the last moment after a long execution. Similarly, to prevent attackers from cheaply clogging the mempool, nodes in the P2P network need to check if an operation will pay a fee before they are willing to forward it.
+
+In this proposal, we expect accounts to have a `validateUserOp` method that takes as input a `UserOperation`, and verify the signature and pay the fee. This method is required to be almost-pure: it is only allowed to access the storage of the account itself, cannot use environment opcodes (eg. `TIMESTAMP`), and can only edit the storage of the account, and can also send out ETH (needed to pay the entry point). The method is gas-limited by the `verificationGasLimit` of the `UserOperation`; nodes can choose to reject operations whose `verificationGasLimit` is too high. These restrictions allow block builders and network nodes to simulate the verification step locally, and be confident that the result will match the result when the operation actually gets included into a block.
```

Comment:
`almost pure` - shouldn't we say something like "restricted in compliance with ERC-7562" instead of vaguely saying "almost pure"?
`is required` - I would say "required" is a bit too strong as we have specified ways to make exceptions. Maybe "is expected" or even "should"?

Reply:
Also, this is really not rationale, but a semi deep dive into the technical details of validation rules and gas calculations. Not sure this is the place to do it.